kalite/coachreports/static/js/coachreports/coach_reports/views.js

```diff
@@ -0,0 +1,343 @@
+        _.bindAll(this);
+        this.set_data_model();
+        this.listenTo(this.model, "change", this.set_data_model);
+    },
+
+    render: function() {
+        this.$el.html(this.template({
+            contents: this.contents.toJSON(),
+            learners: this.contents.length
+        }));
+        var row_views = [];
+        for (var i = 0; i < this.learners.length; i++) {
+            var row_view = this.add_subview(TabularReportRowView, {model: this.learners.at(i), contents: this.contents})
+            row_views.push(row_view);
+            this.listenTo(row_view, "detail_view", this.set_detail_view);
+        }
+        this.append_views(row_views, ".student-data");
+    },
+
+    set_data_model: function (){
+        var self = this;
+        this.data_model = new CoachReportModel({
+            facility: this.model.get("facility"),
+            group: this.model.get("group")
+        });
+        if (this.model.get("facility")) {
+            this.data_model.fetch().then(function() {
+                self.learners = new Backbone.Collection(self.data_model.get("learners"));
+                self.contents = new Backbone.Collection(self.data_model.get("contents"));
+                for (var i = 0; i < self.learners.length; i++) {
+                    self.learners.models[i].set("logs", _.object(
+                        _.map(_.filter(self.data_model.get("logs"), function(log) {
+                            return log.user === self.learners.models[i].get("pk");
+                        }), function(item) {
+                            return [item.exercise_id || item.video_id || item.content_id, item];
+                        })));
+                }
+                self.render();
+            });
+        }
+    },
+
+    set_detail_view: function(detail_view) {
+        if (this.detail_view) {
+            this.detail_view.remove();
+        }
+        if (detail_view) {
+            this.detail_view = detail_view;
+        }
+    }
+
+});
+
+var TabularReportRowView = BaseView.extend({
+
+    template: HB.template("tabular_reports/tabular-view-row"),
+
+    tagName: 'tr',
+
+    initialize: function(options) {
+        _.bindAll(this);
+
+        this.contents = options.contents;
+        this.render();
+    },
+
+    render: function() {
+        this.$el.html(this.template(this.model.attributes));
+
+        var cell_views = [];
+        for (var i = 0; i < this.contents.length; i++) {
+            var data = this.model.get("logs")[this.contents.at(i).get("id")];
+            var new_view = this.add_subview(TabularReportRowCellView, {model: new Backbone.Model(data)});
+            cell_views.push(new_view);
+            this.listenTo(new_view, "detail_view", this.show_detail_view);
+        }
+        this.append_views(cell_views);
+    },
+
+    show_detail_view: function(model) {
+        if (this.detail_view) {
+            // TODO (rtibbles): Implement Models properly here to reflect server side id attributes
```

Comment:
I am referencing .cid here for the model client id (i.e. the id of the model as set up on the client).

When you define Backbone Models, you can define an ```idAttribute``` property that allows you to set a particular property of the setting object to act as the id that is then used in the Collection.

If we did this with the Django Model id, we could then dynamically update these models on the client side when they changed on the server, by only pushing up the items that had changed. Because they would then be matched on their id property, the Collection could intelligently update the models that changed.

This would then allow us to set up push notifications to the client, making the coach report interface seamlessly realtime.